TST: streamline orbits tests

jklenzing opened the pull request:
# Description

Addresses #749, progress on #545, #565, #860

- Reconfigures classes in test_orbits.py for better modularity
  - Two basic inheritable classes: general tests and gappy data tests
  - Reduced length of gappy tests - iterating over one week rather than one month
- Improves style and reduces code duplication
  - usage of parametrize
  - `filter_data`functions combined and generalized
  - local functions renamed to avoid confusion with generalized `filter_data`
- Docstrings for test_orbits.py
- Some duplicate tests removed
- Added ignore D200 and D202 to address docstring linting (#860)

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

tested via pytest.

**Test Configuration**:
* Operating system: Mac OS X 10.15.7
* Version number: Python 3.8.2

# Checklist:

- [x] Make sure you are merging into the ``develop`` (not ``main``) branch
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] Add a note to ``CHANGELOG.md``, summarizing the changes

If this is a release PR, replace the first item of the above checklist with the release
checklist on the wiki: https://github.com/pysat/pysat/wiki/Checklist-for-Release

jklenzing commented:
Potentially room for improvement.  This section needs discussion for other tests that can be combined, as well as making sure all use cases are still covered.  Some tests I still don't understand, the docstrings reflect this.  I don't particularly see the rationale for testing for longer gaps than two days, so the five-day gap test has been removed.

Run time down to 30 min.

jklenzing commented on pysat/tests/test_orbits.py:
```diff
@@ -405,160 +434,38 @@ def filter_data(inst):
-            self.testInst.orbits.next()
-        assert all(control.data == self.testInst.data)
-
-    def test_repeated_orbit_calls_antisymmetric_multi_multi_day_off_0_UT(self):
-        """ Test successful orbit calls for more days about a time off 0 UT
-        """
-        self.stime -= dt.timedelta(days=1)
-        self.testInst.load(date=self.stime)
-        self.testInst.orbits.next()
```
This test (and similar) can be combined with an extra level of parametrize, but I'm not sure it adds anything new

rstoneback commented:
Taking a look now.

rstoneback commented on pysat/tests/test_orbits.py:
```diff
@@ -12,7 +12,25 @@
+    """Remove data from instrument, simulating gaps in the dataset"""
+
+    if times is None:
+        times = [[dt.datetime(2009, 1, 1, 1), dt.datetime(2009, 1, 1, 2)],
+                 [dt.datetime(2009, 1, 1, 10), dt.datetime(2009, 1, 1, 12)],
+                 [dt.datetime(2009, 1, 1, 22), dt.datetime(2009, 1, 2, 2)],
+                 [dt.datetime(2009, 1, 4), dt.datetime(2009, 1, 6)]
+                 ]
```
Some specific edge cases are getting lost.
- single orbit gap
- orbit < gap < 2 orbits
- gap across daybreak, gap > 2 orbits, , no extra data for last orbit on 1st, no data in first orbit time on 2nd
- gap > 1 day, data in first orbit time on 15th, no data on 14th
- gap > 1 day, only a little bit of last orbit on 20th and a little bit of first orbit on 25th available.
- orbit < gap < 2 orbits, daybreak check, 
```
    times = [[dt.datetime(2009, 1, 1, 1, 37), dt.datetime(2009, 1, 1, 3, 14)],
              [dt.datetime(2009, 1, 1, 10), dt.datetime(2009, 1, 1, 12)],
              [dt.datetime(2009, 1, 1, 22), dt.datetime(2009, 1, 2, 2)],
              [dt.datetime(2009, 1, 13), dt.datetime(2009, 1, 15)],
              [dt.datetime(2009, 1, 20, 1), dt.datetime(2009, 1, 25, 23)],
              [dt.datetime(2009, 1, 25, 23, 30), dt.datetime(2009, 1, 26, 3)]
```

jklenzing replied:
Do we need a month of data here, or can we compress into a week and capture all these use cases?

rstoneback commented on pysat/tests/test_orbits.py:
```diff
@@ -753,8 +650,41 @@ def teardown(self):
+    def test_repeat_orbit_calls_asym_multi_day_0_UT_really_long_time_gap(self):
+        """Test successful orbit calls over a series of lengthening gaps
+        """
+        self.testInst.load(date=self.stime)
```
The loaded data and the number of orbits determines which gap we run into. Based on my list of times, we should load this one on the 19th and go for like 45 orbits.

jklenzing replied:
@rstoneback, I've added the check for where the orbit breaks are here, and it is failing for the 45 orbit test on the 19th.  We could drop these added tests for now and add an issue.

jklenzing replied:
UPDATE: Since this is a new test, I've dropped the check and added #861 to document

rstoneback commented on pysat/tests/test_orbits.py:
```diff
@@ -753,8 +650,41 @@ def teardown(self):
+        control = self.testInst.copy()
+        n_time = []
+        p_time = []
+        for j in range(40):
```
20 orbits ok for this one with a load on 2009, 1

rstoneback replied:
Also would be good to load on the 25th and go for like 20 orbits.

rstoneback commented on pysat/tests/test_orbits.py:
```diff
@@ -769,7 +699,10 @@ def teardown(self):
 
 
-class TestOrbitsGappyData2(TestGeneralOrbitsMLT):
+class TestOrbitsGappyData2(TestOrbitsGappyData):
```
I think the gappydata2 class came later after a bug was found. I don't think it should be running the TestOrbitsGappyData tests.

This class is checking if the first orbit is missing all data, and the first orbit sample starts on the day. Also has a single orbit gap unrelated to daybreak. I think we just need to do a standard iterate forward 20 orbits starting from 2009, 1, back 40, forward 20 type thing here.

jklenzing commented:
Changes since last review:
- Using recommended time gaps with targeted tests
- merged in recent changes from develop
- scrubbed docstrings and usage of return

The tests here take between 21 and 37 minutes on GA (unix and windows)
In develop, it takes 50 to 67 minutes.

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -12,26 +12,48 @@
 
 
+def filter_data(inst, times=None):
+    """Remove data from instrument, simulating gaps in the dataset."""
```
```suggestion
    """Remove data from instrument, simulating gaps in the data set."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -12,26 +12,48 @@
+def filter_data(inst, times=None):
+    """Remove data from instrument, simulating gaps in the dataset."""
+
+    if times is None:
```
Recommend making this more robust by using the instrument index start time and adding different deltas to it.

jklenzing replied:
In order to import the instrument test date, I had to move this into the class for `TestOrbitsGappyData`.

jklenzing replied:
Updated to use `setup_class` and `teardown_class`

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -12,26 +12,48 @@
         self.in_args = ['pysat', 'testing']
         self.in_kwargs = {'clean_level': 'clean', 'update_files': True}
         self.testInst = None
         self.stime = dt.datetime(2009, 1, 1)
```
Use instrument testing time

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -12,26 +12,48 @@
-        """
+        """Test orbit failure with bad 'kind' input."""
         self.in_kwargs['orbit_info'] = {'index': 'mlt', 'kind': 'cats'}
         with pytest.raises(ValueError):
```
Add an error message evaluation.

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -43,14 +65,14 @@ def test_orbit_w_bad_kind(self):
         self.testInst = pysat.Instrument(*self.in_args, **self.in_kwargs)
         self.testInst.load(date=self.stime)
 
         with pytest.raises(ValueError):
```
Add an error message evaluation

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -61,8 +83,7 @@ def test_orbit_w_bad_orbit_info(self, info):
     def test_orbit_polar_w_missing_orbit_index(self, info):
-        """ Test orbit failure oon iteration with missing orbit index
-        """
+        """Test orbit failure oon iteration with missing orbit index."""
```
```suggestion
        """Test orbit failure on iteration with missing orbit index."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -71,49 +92,51 @@ def test_orbit_polar_w_missing_orbit_index(self, info):
         self.testInst.load(date=self.stime)
         with pytest.raises(ValueError):
```
Add an error message evaluation.

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -174,55 +197,57 @@ def test_all_single_orbit_calls_in_day(self):
-        """
+        """Test orbit next call without loading data."""
         self.testInst.orbits.next()
         assert (self.testInst.index[0] == dt.datetime(2008, 1, 1))
         assert (self.testInst.index[-1] == dt.datetime(2008, 1, 1, 0, 38, 59))
```
Should these times be related back to one of the test class attributes?

aburrell replied:
Here and elsewhere.

jklenzing replied:
I think this is all hard wired.  @rstoneback?

jklenzing replied:
This test and the next one are confusing me.  I get that if no data is loaded it loops around to the last file.  But why is the orbit end 38 minutes here and 11 minutes for the other test?  If we change the test date down the line, are these numbers going to be the same, or will it find different break points?

aburrell replied:
Did you create an issue for these fixed date tests?

rstoneback replied:
The test date is set on when MLT rolls-over from 24 back to 0, plot attached. 
![orbits_mlt](https://user-images.githubusercontent.com/3595641/128202288-b1832a8f-f17a-42dd-aa7a-3c5ff0e226b9.png)

No, if the test date is changed it won't be the same. The correct times can be calculated using the known period and root_date of the 'met' signal.

rstoneback replied:
The start time check is driven by the fact that it is the first file loaded (last file in following test). The file dates in self.testInst.files.files.index could be used for a more generalized approach.

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -316,69 +344,56 @@ def test_eval_repr(self):
         self.testInst.orbits.next()
 
-        # a recusion issue has been observed in this area
+        # a recursion issue has been observed in this area
```
```suggestion
        # A recursion issue has been observed in this area
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -316,69 +344,56 @@ def test_eval_repr(self):
 
-        # a recusion issue has been observed in this area
+        # a recursion issue has been observed in this area
         # checking for date to limit reintroduction potential
```
```suggestion
        # Checking for date to limit reintroduction potential
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -316,69 +344,56 @@ def test_eval_repr(self):
-        self.testInst.load(date=self.stime)
-        # starting from no orbit calls next loads first orbit
-        self.testInst.orbits.next()
         # store comparison data
```
```suggestion

        # Store comparison data
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -403,181 +418,74 @@ def filter_data(inst):
             assert all(self.testInst.data == saved_data.data)
 
-        # a recusion issue has been observed in this area
+        # a recursion issue has been observed in this area
```
```suggestion
        # A recursion issue has been observed in this area.
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -403,181 +418,74 @@ def filter_data(inst):
 
-        # a recusion issue has been observed in this area
+        # a recursion issue has been observed in this area
         # checking for date to limit reintroduction potential
```
```suggestion
        # Checking for date to limit reintroduction potential.
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -751,13 +658,49 @@ def setup(self):
+
+    def test_repeat_orbit_calls_sym_multi_day_0_UT_really_long_time_gap(self):
+        """Test successful orbit calls over a multi-day gap."""
+        self.testInst.load(date=dt.datetime(2009, 1, 19))
```
Set year, month from `self.stime`

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -751,13 +658,49 @@ def setup(self):
+    @pytest.mark.parametrize("day", [1, 25])
+    def test_repeat_orbit_calls_cutoffs_with_gaps(self, day):
+        """Test that orbits are selected at same cutoffs when reversed."""
+        self.testInst.load(date=dt.datetime(2009, 1, day))
```
Set year, month from self.stime

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -751,13 +658,49 @@ def setup(self):
+            self.testInst.orbits.prev()
+            p_time.append(self.testInst.index[0])
+
+        check = np.all(p_time == n_time[::-1])
+        assert all(control.data == self.testInst.data) & check
```
Check one thing at a time for clear assertion errors

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -817,14 +782,18 @@ def setup(self):
 
-class TestOrbitsGappyLongData(TestGeneralOrbitsMLT):
+class TestOrbitsGappyLongData(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Longitude -- pandas"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Longitude -- pandas."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -834,13 +803,17 @@ def setup(self):
 
-class TestOrbitsGappyLongDataXarray(TestGeneralOrbitsMLT):
+class TestOrbitsGappyLongDataXarray(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Longitude -- xarray"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Longitude -- xarray."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -849,13 +822,17 @@ def setup(self):
+
 
+class TestOrbitsGappyOrbitNumData(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Orbit Number -- pandas"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Orbit Number -- pandas."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -864,13 +841,17 @@ def setup(self):
 
-class TestOrbitsGappyOrbitNumDataXarray(TestGeneralOrbitsMLT):
+class TestOrbitsGappyOrbitNumDataXarray(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Orbit Number -- xarray"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Orbit Number -- xarray."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -879,13 +860,17 @@ def setup(self):
+
 
+class TestOrbitsGappyOrbitLatData(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Latitude -- pandas"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Latitude -- pandas."""
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -895,13 +880,17 @@ def setup(self):
 
-class TestOrbitsGappyOrbitLatDataXarray(TestGeneralOrbitsMLT):
+class TestOrbitsGappyOrbitLatDataXarray(TestOrbitsGappyData):
+    """Run the gappy orbit tests for orbits defined by Latitude -- xarray"""
```
```suggestion
    """Run the gappy orbit tests for orbits defined by Latitude -- xarray."""
```

aburrell reviewed (commented):
Block comment grammar needs to be fixed throughout.

jklenzing commented:
Last iteration uses `flake8-docstrings` to scan docstrings for consistency.
```
flake8 . --count --select=D,E,F,W
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -1,3 +1,5 @@
+"""Test the pysat routines for the orbits class."""
```
Recommend adding the license info to the head of the file (check one of the recent tests I updated).

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -12,47 +14,70 @@
 import pysat
 
 
+def filter_data(inst, times=None):
```
```suggestion
def filter_data(inst, times):
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -316,82 +349,69 @@ def test_eval_repr(self):
 
-        # a recusion issue has been observed in this area
-        # checking for date to limit reintroduction potential
+        # A recursion issue has been observed in this area
+        # Checking for date to limit reintroduction potential
```
```suggestion
        # A recursion issue has been observed in this area.
        # Checking for date to limit reintroduction potential
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -403,183 +423,76 @@ def filter_data(inst):
 
-        # a recusion issue has been observed in this area
-        # checking for date to limit reintroduction potential
+        # A recursion issue has been observed in this area
+        # Checking for date to limit reintroduction potential
```
```suggestion
        # A recursion issue has been observed in this area.
        # Checking for date to limit reintroduction potential
```

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -403,183 +423,76 @@ def filter_data(inst):
+        for j in range(iterations):
+            n_time.append(self.testInst.index[0])
             self.testInst.orbits.next()
-        for j in range(40):
+        for j in range(iterations):
             self.testInst.orbits.prev()
-        for j in range(20):
-            self.testInst.orbits.next()
+            p_time.append(self.testInst.index[0])
```
This "reversability" code snippit is repeated a lot.  It could probably be an evaluation method in the core class.

aburrell reviewed (commented):
More comments, though some from my last review are still waiting for input from @rstoneback

aburrell commented on pysat/tests/test_orbits.py:
```diff
@@ -316,82 +436,73 @@ def test_eval_repr(self):
         if self.testInst.orbits.num == 1:
-            # equivalence only when only one orbit
-            # some test settings can violate this assumption
+            # Equivalence only when only one orbit
+            # Some test settings can violate this assumption
```
```suggestion
            # Equivalence occurs only when there is one orbit,
            # some test settings can violate this assumption.
```

aburrell reviewed (commented):
Looking good, found one small thing.  Want to verify that the hard-coded dates are contained in an issue as well.

jklenzing commented:
Added #863 to capture this discussion.
